File: backtracking/leetcode-22-generate-parentheses.java
// Leetcode 22. Generate Parentheses
//
// Given n pairs of parentheses, write a function to generate all combinations of well-formed parentheses.
//
// For example, given n = 3, a solution set is:
//
// [
//   "((()))",
//   "(()())",
//   "(())()",
//   "()(())",
//   "()()()"
// ]
//

class Solution {
    public List<String> generateParenthesis(int n) {
        List<String> results = new ArrayList();
        backtrack("", 0, 0, results, n);

        return results;
    }

    private void backtrack(String current, int open, int close, List<String> results, int max) {
        if (current.length() == 2 * max) {
            results.add(current);
            return;
        }

        if (open < max) {
            backtrack(current + "(", open + 1, close, results, max);
        }

        if (close < open) {
            backtrack(current + ")", open, close + 1, results, max);
        }
    }
}

File: backtracking/leetcode-22-generate-parentheses.rb
# Leetcode 22. Generate Parentheses
#
# Given n pairs of parentheses, write a function to generate all combinations of well-formed parentheses.
#
# For example, given n = 3, a solution set is:
#
# [
#   "((()))",
#   "(()())",
#   "(())()",
#   "()(())",
#   "()()()"
# ]
#

# @param {Integer} n
# @return {String[]}
def generate_parenthesis(n)
    results = []
    backtrack("", 0, 0, results, n)

    results
end

def backtrack(current, open, close, results, max)
    if current.length == max * 2
        results.push(current)
        return
    end

    if open < max
        backtrack(current + "(", open + 1, close, results, max)
    end

    if close < open
        backtrack(current + ")", open, close + 1, results, max)
    end
end
